### PR TITLE
Introduce StaticWhisperPipeline

### PR DESCRIPTION
**Pull request by TolyaTalamanov**

### CODE REVIEW

File: samples/cpp/whisper_speech_recognition/whisper_speech_recognition.cpp [hunk 1]
```diff
@@ -5,16 +5,17 @@
 #include "openvino/genai/whisper_pipeline.hpp"
 
 int main(int argc, char* argv[]) try {
-    if (3 > argc) {
-        throw std::runtime_error(std::string{"Usage: "} + argv[0] + " <MODEL_DIR> \"<WAV_FILE_PATH>\"");
+    if (4 > argc) {
```
Comment (TolyaTalamanov) [inline samples/cpp/whisper_speech_recognition/whisper_speech_recognition.cpp:8]: Will return back

File: src/cpp/include/openvino/genai/whisper_pipeline.hpp [hunk 2]
```diff
@@ -17,9 +17,10 @@ using OptionalWhisperGenerationConfig = std::optional<WhisperGenerationConfig>;
 
 using RawSpeechInput = std::vector<float>;
 
+class WhisperPipelineImplBase;
```
Comment (ilya-lavrenov) [inline src/cpp/include/openvino/genai/whisper_pipeline.hpp:20]: could you please hide it inside private section of WhisperPipeline as it's done for Text2Image ?

https://github.com/openvinotoolkit/openvino.genai/blob/abbc695715a1d0a92facd807fe63c6d2088b99e7/src/cpp/include/openvino/genai/text2image/pipeline.hpp#L144-L151

File: src/cpp/src/whisper_pipeline_base.hpp [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    : m_generation_config(config), m_tokenizer(tokenizer), m_feature_extractor(feature_extractor) {
+    }
+
+    virtual DecodedResults generate(const RawSpeechInput& raw_speech_input,
+                                    OptionalWhisperGenerationConfig generation_config,
+                                    StreamerVariant streamer) = 0;
+
+    virtual ~WhisperPipelineImplBase() = default;
+
+    WhisperGenerationConfig m_generation_config;
```
Comment (ilya-lavrenov) [inline src/cpp/src/whisper_pipeline_base.hpp:28]: move to protected section all the fields?

File: src/cpp/src/whisper_pipeline.cpp [hunk 4]
```diff
@@ -41,21 +45,17 @@ std::vector<int64_t> whisper_generate(const ov::genai::WhisperGenerationConfig&
                                       ov::genai::WhisperFeatureExtractor& feature_extractor,
                                       const std::shared_ptr<StreamerBase> streamer);
 
-class WhisperPipeline::Impl {
+class WhisperPipelineImpl : public WhisperPipelineImplBase {
```
Comment (ilya-lavrenov) [inline src/cpp/src/whisper_pipeline.cpp:48]:
```suggestion
class WhisperPipelineStatefulImpl : public WhisperPipelineImplBase {
```

And I would separate `WhisperPipelineStatefulImpl` and `WhisperPipeline` classes to different source files.

File: src/cpp/src/whisper_pipeline.cpp [hunk 5]
```diff
@@ -105,7 +105,11 @@ ov::genai::WhisperPipeline::WhisperPipeline(const std::string& model_path,
                                             const std::string& device,
                                             const ov::AnyMap& plugin_config) {
     auto start_time = std::chrono::steady_clock::now();
-    m_impl = std::make_unique<WhisperPipeline::Impl>(model_path, tokenizer, device, plugin_config);
+    if (device == "NPU") {
```
Comment (ilya-lavrenov) [inline src/cpp/src/whisper_pipeline.cpp:108]: can device be `NPU.<device ID>` ? Looks like we need to detect whether device name starts with `NPU`

File: src/cpp/include/openvino/genai/whisper_pipeline.hpp (hunk 2 above)
Comment (TolyaTalamanov) [reply src/cpp/include/openvino/genai/whisper_pipeline.hpp:20]: I don't think so, it will prohibit inheritance from this class. I believe this is the reason why it's done the same way for `LLMPipelineImplBase`:
https://github.com/openvinotoolkit/openvino.genai/blob/abbc695715a1d0a92facd807fe63c6d2088b99e7/src/cpp/include/openvino/genai/llm_pipeline.hpp#L99-L115

File: src/cpp/src/whisper_pipeline_base.hpp (hunk 3 above)
Comment (TolyaTalamanov) [reply src/cpp/src/whisper_pipeline_base.hpp:28]: Fields are being accessed like this: https://github.com/openvinotoolkit/openvino.genai/blob/abbc695715a1d0a92facd807fe63c6d2088b99e7/src/cpp/src/whisper_pipeline.cpp#L114

In that case, it might require introducing "getters". I kept it like this to minimize the changes.

File: src/cpp/src/whisper_pipeline.cpp (hunk 5 above)
Comment (TolyaTalamanov) [reply src/cpp/src/whisper_pipeline.cpp:108]: `NPU.<device ID>` such format does make sense for compilation on the machine w/o NPU.

But for runtime, NPU should exists on the machine therefore architecture will be detected automatically.

P.S It's a good point and might be useful in case user want to precompile blobs via GenAI, but this flow isn't in-place yet.

File: src/cpp/include/openvino/genai/whisper_pipeline.hpp (hunk 2 above)
Comment (ilya-lavrenov) [reply src/cpp/include/openvino/genai/whisper_pipeline.hpp:20]: see how it's done in Text2Image

Inherited class should be a friend of main public class.

File: src/cpp/src/whisper_pipeline.cpp [hunk 6]
```diff
@@ -41,21 +45,17 @@ std::vector<int64_t> whisper_generate(const ov::genai::WhisperGenerationConfig&
                                       ov::genai::WhisperFeatureExtractor& feature_extractor,
                                       const std::shared_ptr<StreamerBase> streamer);
 
-class WhisperPipeline::Impl {
+class WhisperPipelineStatefulImpl : public WhisperPipelineImplBase {
```
Comment (TolyaTalamanov) [inline src/cpp/src/whisper_pipeline.cpp:48]: @ilya-lavrenov is `WhisperPipelineStatefulImpl` correct naming? Models aren't stateful, they expose KV-cache outside

Comment (ilya-lavrenov) [reply src/cpp/src/whisper_pipeline.cpp:48]: good point.

Let's keep name as is and rename to stateful as they really become stateful

File: src/cpp/src/whisper_pipeline_static.cpp [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+int64_t decode(ov::Tensor& encoder_hidden_state,
+               ov::InferRequest& decoder,
+               std::vector<int32_t>& input_ids,
+               const ov::genai::WhisperGenerationConfig& config,
+               bool do_suppress_tokens = true) {
+    // FIXME: Is this input connected with anything???
+    decoder.set_tensor("encoder_hidden_states", ov::Tensor{encoder_hidden_state});
+    ov::Tensor input_ids_tensor(ov::element::i32, { 1, input_ids.size() }, input_ids.data());
+    decoder.set_tensor("input_ids", input_ids_tensor);
+
```
Comment (TolyaTalamanov) [inline src/cpp/src/whisper_pipeline_static.cpp:134]: "attention_mask" should be also filled, but missed here!

Comment (TolyaTalamanov) [reply src/cpp/src/whisper_pipeline_static.cpp:134]: Fixed

File: samples/cpp/whisper_speech_recognition/whisper_speech_recognition.cpp (hunk 1 above)
Comment (TolyaTalamanov) [reply samples/cpp/whisper_speech_recognition/whisper_speech_recognition.cpp]: Done

File: src/cpp/src/whisper_pipeline.cpp [hunk 8]
```diff
@@ -13,6 +13,10 @@
 #include "whisper/whisper_feature_extractor.hpp"
 #include "whisper/whisper_models.hpp"
 
+#include "whisper_pipeline_base.hpp"
+#include "whisper_pipeline_static.hpp"
+
+// FIXME: Remove these duplicates
```
Comment (TolyaTalamanov) [inline src/cpp/src/whisper_pipeline.cpp:19]: Looks can't be simply moved to utils as there are already functions with such signature for llm pipeline

File: src/cpp/include/openvino/genai/whisper_pipeline.hpp (hunk 2 above)
Comment (ilya-lavrenov) [reply src/cpp/include/openvino/genai/whisper_pipeline.hpp:20]: ping

**Review by ilya-lavrenov — APPROVED** [review]

**Comment by soumendukrg** [comment]
Is there any planned timeline for this merge?

File: src/cpp/src/whisper_pipeline_static.hpp
Comment (soumendukrg) [inline src/cpp/src/whisper_pipeline_static.hpp:1]: What does the plugin config signify here?

File: src/cpp/src/whisper_pipeline_static.cpp [hunk 9]
```diff
@@ -0,0 +1,362 @@
...
+        }
+
+        std::string with_past_input_name =
+            std::regex_replace(source_output_name, std::regex("present"), "past");
+
+        auto src_kv_tensor = source.get_tensor(source_output_name);
+        auto dst_kv_tensor = dest.get_tensor(with_past_input_name);
+        auto kv_size = src_kv_tensor.get_shape()[2];
+        // NB: Copy src_kv_tensor into dst_kv_tensor[:, :, kv_pos:kv_pos+kv_size, :]
+        auto dst_kv_tensor_slice = make_tensor_slice(dst_kv_tensor, 2u, kv_pos, kv_pos + kv_size);
```
Comment (soumendukrg) [inline src/cpp/src/whisper_pipeline_static.cpp:128]: For the next optimization, considering optimal layout, seq length (kv_size) should be in dim 2 for keys and dim 3 for values.

File: src/cpp/src/whisper_pipeline_static.cpp [hunk 10]
```diff
@@ -0,0 +1,362 @@
...
+    return output_token;
+}
+
+int64_t decode_with_past(ov::InferRequest& decoder_with_past,
+                         const int64_t input_id,
+                         const int64_t position_id,
+                         const ov::genai::WhisperGenerationConfig& config) {
+
+    // FIXME: Avoid this cast to i32. Why it's not i64 precision in model?
+    decoder_with_past.get_tensor("input_ids").data<int32_t>()[0] = static_cast<int32_t>(input_id);
```
Comment (soumendukrg) [inline src/cpp/src/whisper_pipeline_static.cpp:160]: Tokens will always be representable in i32. i64 is not needed. It is advised to change all input_ids to i32 instead of casting this to i32.

File: src/cpp/src/whisper_pipeline_static.cpp [hunk 11]
```diff
@@ -0,0 +1,362 @@
...
+    // FIXME: Avoid this cast to i32. Why it's not i64 precision in model?
+    decoder_with_past.get_tensor("position_ids").data<int32_t>()[0] = static_cast<int32_t>(position_id);
+    // FIXME: Is "attention_mask" supposed to be f16?
```
Comment (soumendukrg) [inline src/cpp/src/whisper_pipeline_static.cpp:163]: Mask is used later for computation. If mask is already in FP16, this avoids extra conversion step on device.

File: src/cpp/src/whisper_pipeline_static.cpp [hunk 12]
```diff
@@ -0,0 +1,362 @@
...
+}
+
+int64_t decode_with_past(ov::InferRequest& decoder_with_past,
+                         const int64_t input_id,
+                         const int64_t position_id,
+                         const ov::genai::WhisperGenerationConfig& config) {
+
+    // FIXME: Avoid this cast to i32. Why it's not i64 precision in model?
+    decoder_with_past.get_tensor("input_ids").data<int32_t>()[0] = static_cast<int32_t>(input_id);
+    // FIXME: Avoid this cast to i32. Why it's not i64 precision in model?
```
Comment (soumendukrg) [inline src/cpp/src/whisper_pipeline_static.cpp:161]: Position ids can max go upto 448 for whisper. I32 is sufficient.

File: src/cpp/src/whisper_pipeline_static.cpp (hunk 10 above)
Comment (TolyaTalamanov) [reply src/cpp/src/whisper_pipeline_static.cpp:160]: `input_ids` cannot be changed to `i32`, as tokenizer is common for `static` and `stateful` pipeline

File: src/cpp/src/whisper_pipeline_static.cpp (hunk 12 above)
Comment (TolyaTalamanov) [reply src/cpp/src/whisper_pipeline_static.cpp:161]: In more generic case, kvcache size might be higher, but ofc will fit into `i32`

Comment (soumendukrg) [reply src/cpp/src/whisper_pipeline_static.cpp:161]: Yes, in a generic case, kv cache sequence length can become a very large value. Optimum-cli generated model use i64 precision for position ids (name is now changed to cache_position).
